rspace/src/main/scala/coop/rchain/rspace/internal.scala

```diff
@@ -36,6 +39,7 @@ object internal {
         patterns: Seq[P],
         continuation: K,
         persist: Boolean,
+        peek: SortedSet[Int],
```

Comment:
I think it'd make sense to have the Channels be a `Seq[(C, Boolean)]`, or a Seq[ChannelRead], where ChannelRead ~ (C, Boolean)

Reply:
cost is: we have to encode the tuple
what's the gain?

Reply:
We have the info about the channels on one field vs two. We're explicit in saying whether a given channel is a peek-channel, and don't rely on indexes that could be de-sync from the indexes of the channels.

Speaking of which: are there places where we sort the channels list? If yes, do the peek indices get mapped through the resulting permutation?

Reply:
I don't follow. why is it that `patterns` is fine being extracted as a separate parameter (which is dense, every channel has to have a pattern) and peek (which is sparse) has to be part of the channels structure? this seems counter intuitive to me.

we are not sorting the channels as they are basis for the key in looking up a continuation.

Reply:
You have a point that the pattern should also be a part of a `readChannel / bindPattern / channelRead / whatever we call it`, and I :100: agree with you on that ;)

re: sparse vs dense - this matters mostly during serialization, which is (luckily) decoupled from the ADT - at least in rspace.

I'm just suggesting a data representation that avoids indexing and instead aggregates data that constitute one whole in tuples/products, to make invalid states irrepresentable (different number of patterns  than there are channels in the consume / elements in the `peek` set that are not indexes in patterns / source.channels).

I guess though when we add this all this is a big refactoring effort, so let's just ponder it for a while and maybe consider next time. I hope the fields of WaitingContinuation now irk you just as much as they irk me :)

Reply:
ah! so it's a bigger point! I started with that approach (there is a closed pr with that somewhere) and it's a mess.